common/journald/src/stream.rs

```diff
@@ -0,0 +1,356 @@
+                        e
+                    );
+                    break;
+                }
+            }
+
+            // some sort of error has occurred. Explicitly drop the sender before waking up the
+            // stream to prevent a race condition
+            drop(sender);
+            call_waker();
+        });
+
+        self.thread = Some(thread);
+        self.receiver = Some(receiver);
+    }
+
+    fn drop_thread(&mut self) {
+        self.shared_state.lock().unwrap().is_alive = false;
+        if let Some(thread) = self.thread.take() {
+            if let Err(e) = thread.join() {
+                warn!("unable to join journald's worker thread: {:?}", e)
+            }
+        }
+    }
+}
+
+impl FutureStream for Stream {
+    type Item = Vec<LineBuilder>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
+        let mut self_ = self.as_mut();
+
+        if let Some(ref receiver) = self_.receiver {
+            match receiver.try_recv() {
+                Ok(line) => {
+                    return Poll::Ready(Some(vec![line]));
+                }
+                Err(TryRecvError::Disconnected) => {
+                    warn!("journald's main thread unable to read from worker thread, restarting worker thread...");
+                    self_.drop_thread();
+                    self_.spawn_thread();
+                }
+                _ => {}
+            }
+        } else {
+            warn!(
+                "journald's main thread missing connection to worker thread, shutting down stream"
+            );
+            return Poll::Ready(None);
+        }
+
+        let mut shared_state = self_.shared_state.lock().unwrap();
+        shared_state.waker = Some(cx.waker().clone());
+        Poll::Pending
+    }
+}
+
+impl Drop for Stream {
+    fn drop(&mut self) {
+        self.drop_thread();
+    }
+}
+
+struct Reader {
+    reader: Journal,
+}
+
+impl Reader {
+    fn new(path: Path) -> Self {
+        let mut reader = match path {
+            Path::Directory(path) => Journal::open_directory(&path, JournalFiles::All, false)
+                .expect("Could not open journald reader for directory"),
+            Path::Files(paths) => {
+                let paths: Vec<&std::path::Path> = paths.iter().map(PathBuf::as_path).collect();
+                Journal::open_files(&paths).expect("Could not open journald reader for paths")
+            }
+        };
+        reader
+            .seek(JournalSeek::Tail)
+            .expect("Could not seek to tail of journald logs");
+
+        Self { reader }
+    }
+
+    fn process_next_record(&mut self) -> RecordStatus {
+        let record = match self.reader.next_entry() {
+            Ok(Some(record)) => record,
+            Ok(None) => return RecordStatus::NoLines,
+            Err(e) => panic!("Unable to read next record from journald: {}", e),
```

Comment:
Is this condition unrecoverable? Or could it arise if the journald daemon was restarted or something? In which case I think it would be better return a Result here and let the worker thread try recreating it's journal object before resorting to the infinite retry logic in the Stream? (which feels like exception handling and a bit un-rustic)